### PR TITLE
Remove a stray fmt.Printf, use a logger

### PR DESCRIPTION
**Pull request by atoulme**
**Description:** 
Remove a stray fmt.Printf, use a logger

**Link to tracking Issue:** 
#14467

**Testing:**
N/A

**Documentation:**
N/A

### CODE REVIEW

**Review by kentquirk — CHANGES_REQUESTED** [review]
I'm home with COVID at the moment and not in a position to do any performance analysis today.

But I'm not very happy about adding a logger parameter everywhere and passing it on every function call. This code is designed to be as fast as possible.

The printf you're complaining about iin #14467 is telling you that you gave it bad data -- the example code in the bug (please link the bug with # before the bug number in the original so that GitHub does the right things with it!) is telling you that `attributes["foo"]` is not a string but you're comparing it to one.

In the original version of this code, I didn't have the printf at all, and then was asked to include it because we should generate some. 

I guess if we think we need it, I would much prefer to move all of the functions here into member functions on the parser.  There are also benchmarks in compare_test.go that I'd want to run a before/after comparison with, whichever path we take.

And finally, I don't believe that either transformprocessor or ottl has introduced a logger before. If we do, I don't think I want to depend on zap as the standard logging interface; I would prefer a more generic choice, especially now that there's some consideration going on in the Go community towards a more flexible standard logging model. Given that this is just a single call, I'd suggest we start with the Go standard log interface, or even define a `logger` interface locally that we can tweak over time.

@TylerHelmuth do you have an opinion on any of this?

**Comment by TylerHelmuth** [comment]
- @kentquirk I agree with everything you said except for not depending on zap. When @evan-bradley introduced the generic logger framework that was definitely the correct solution, but since that decision was made we've pivoted OTTL away from an agnostic solution and steered it towards a collector-specific solution. 

With the pivot in mind, I think zap makes sense since the collector's logger of choice is zap. The OTTL parser now requires the collector's TelemetrySettings struct, which includes its zap logger. (You we're out for this [PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14389) but still willing to discuss if you have thoughts).

**Comment by TylerHelmuth** [comment]
I would also say we do want to replace the print statement with the logger usage so it can conform with the collectors logging settings. I agree that I would rather not pass it around.

**Comment by kentquirk** [comment]
Ok, thanks. I agree that if the collector has standardized on zap that we can use that (although I wish it had been a more general solution, it's done now).

So then, the right answer here is, I think, to move the compare functions into private member functions so that we don't have to add the logger to their signatures.

**Comment by atoulme** [comment]
> I'm home with COVID at the moment and not in a position to do any performance analysis today.
Please rest up!
> 
> But I'm not very happy about adding a logger parameter everywhere and passing it on every function call. This code is designed to be as fast as possible.
It would be best not to print out anything then.
> 
> The printf you're complaining about iin #14467 is telling you that you gave it bad data -- the example code in the bug (please link the bug with # before the bug number in the original so that GitHub does the right things with it!) is telling you that `attributes["foo"]` is not a string but you're comparing it to one.
This is not bad data - attributes["foo"] is a nil value, and this new behavior now makes us print this line for every record, which is a regression of our system. We have customers asking for a fix at once.

There is a simpler fix - we delete the fmt.Printf line.

**Comment by TylerHelmuth** [comment]
This print statement is serving a useful purpose of letting customers know when there are issues or unexpected outcomes during hot path execution. The issue this print statement is informing users about cannot be determined during collector validation. For this reason we need to continue to inform users of this situation. 

This print statement, as well as other situations we haven't addressed yet in Context and Functions, should use the collectors logger so that the processor honors the collectors telemetry settings for logging.

**Comment by dmitryax** [comment]
Just to summarize what discussion here and in the issue.

We all agree that the log message is valid, but should be replaced with a debug log entry instead.

This PR reaches the goal, but passing the logger around as an argument is not ideal solution, so we want to move the functions to member functions on the parser. 

@TylerHelmuth @kentquirk please correct me if I'm missing anything.

My question if we can merge this PR as is and handle the refactoring mentioned above as part of another issue. @TylerHelmuth @kentquirk WDYT?

**Comment by TylerHelmuth** [comment]
@dmitryax you've got it all summarized correctly.  

I'd really rather not fix https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14467 via passing the logger around as it is a fix where we are instantly introducing more work for ourselves and potentially degrading performance (we have benchmarks for comparison to test this).

If we can't fix the issue using member functions at this time and absolutely need to deal with the spurious logs before the next release I'd rather the print statement be removed entirely and a new, high priority issue opened to address passing logging around the parser.  This isn't ideal, but it feels better than adding code we already know we want to remove that could affect performance.

@atoulme if you can run the comparison benchmarks in the package on the main branch and on this PR's branch and confirm there is no performance degradation, then I'd be more inclined to handled the issue via passing the logger.

**Comment by atoulme** [comment]
Here you go, nice little performance improvement as zap doesn't block like `fmt` does.


goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkCompareEQInt64
BenchmarkCompareEQInt64-16       	206829811	         5.394 ns/op
BenchmarkCompareEQFloat
BenchmarkCompareEQFloat-16       	235811641	         5.546 ns/op
BenchmarkCompareEQString
BenchmarkCompareEQString-16      	100000000	        10.49 ns/op
BenchmarkCompareEQPString
BenchmarkCompareEQPString-16     	253092255	         4.790 ns/op
BenchmarkCompareEQBytes
BenchmarkCompareEQBytes-16       	131612401	         9.184 ns/op
BenchmarkCompareEQNil
BenchmarkCompareEQNil-16         	501326972	         2.349 ns/op
BenchmarkCompareNEInt
BenchmarkCompareNEInt-16         	210399183	         6.153 ns/op
BenchmarkCompareNEFloat
BenchmarkCompareNEFloat-16       	200999788	         5.971 ns/op
BenchmarkCompareNEString
BenchmarkCompareNEString-16      	100000000	        10.24 ns/op
BenchmarkCompareLTFloat
BenchmarkCompareLTFloat-16       	227600818	         5.554 ns/op
BenchmarkCompareLTString
BenchmarkCompareLTString-16      	120098608	        10.34 ns/op
BenchmarkCompareLTNil
BenchmarkCompareLTNil-16         	535652295	         2.035 ns/op
BenchmarkCompareEQFunction
BenchmarkCompareEQFunction-16    	167140707	         7.545 ns/op

goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkCompareEQInt64
BenchmarkCompareEQInt64-16       	217768292	         6.375 ns/op
BenchmarkCompareEQFloat
BenchmarkCompareEQFloat-16       	208945070	         5.705 ns/op
BenchmarkCompareEQString
BenchmarkCompareEQString-16      	127988530	         9.537 ns/op
BenchmarkCompareEQPString
BenchmarkCompareEQPString-16     	199986238	         5.941 ns/op
BenchmarkCompareEQBytes
BenchmarkCompareEQBytes-16       	100000000	        10.65 ns/op
BenchmarkCompareEQNil
BenchmarkCompareEQNil-16         	402717524	         3.016 ns/op
BenchmarkCompareNEInt
BenchmarkCompareNEInt-16         	174952662	         6.742 ns/op
BenchmarkCompareNEFloat
BenchmarkCompareNEFloat-16       	197801281	         6.230 ns/op
BenchmarkCompareNEString
BenchmarkCompareNEString-16      	100000000	        10.24 ns/op
BenchmarkCompareLTFloat
BenchmarkCompareLTFloat-16       	189175017	         7.031 ns/op
BenchmarkCompareLTString
BenchmarkCompareLTString-16      	99492043	        11.55 ns/op
BenchmarkCompareLTNil
BenchmarkCompareLTNil-16         	388223613	         2.950 ns/op
BenchmarkCompareEQFunction
BenchmarkCompareEQFunction-16    	143433739	         8.358 ns/op
PASS

Process finished with the exit code 0

File: pkg/ottl/compare.go [hunk 1]
```diff
@@ -16,20 +16,24 @@ package ottl // import "github.com/open-telemetry/opentelemetry-collector-contri
 
 import (
 	"bytes"
-	"fmt"
 
+	"go.uber.org/zap"
 	"golang.org/x/exp/constraints"
 )
 
+type cmp struct {
```
Comment (TylerHelmuth) [inline pkg/ottl/compare.go:24]: Instead of a new type, can we add these to `Parser`?

**Review by TylerHelmuth — COMMENTED** [review]
@atoulme thanks for running those benchmarks

File: pkg/ottl/compare.go (hunk 1 above)
Comment (atoulme) [reply pkg/ottl/compare.go]: done.

**Comment by atoulme** [comment]
New benchmark:

goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkCompareEQInt64
BenchmarkCompareEQInt64-16       	206829811	         5.394 ns/op
BenchmarkCompareEQFloat
BenchmarkCompareEQFloat-16       	235811641	         5.546 ns/op
BenchmarkCompareEQString
BenchmarkCompareEQString-16      	100000000	        10.49 ns/op
BenchmarkCompareEQPString
BenchmarkCompareEQPString-16     	253092255	         4.790 ns/op
BenchmarkCompareEQBytes
BenchmarkCompareEQBytes-16       	131612401	         9.184 ns/op
BenchmarkCompareEQNil
BenchmarkCompareEQNil-16         	501326972	         2.349 ns/op
BenchmarkCompareNEInt
BenchmarkCompareNEInt-16         	210399183	         6.153 ns/op
BenchmarkCompareNEFloat
BenchmarkCompareNEFloat-16       	200999788	         5.971 ns/op
BenchmarkCompareNEString
BenchmarkCompareNEString-16      	100000000	        10.24 ns/op
BenchmarkCompareLTFloat
BenchmarkCompareLTFloat-16       	227600818	         5.554 ns/op
BenchmarkCompareLTString
BenchmarkCompareLTString-16      	120098608	        10.34 ns/op
BenchmarkCompareLTNil
BenchmarkCompareLTNil-16         	535652295	         2.035 ns/op
BenchmarkCompareEQFunction
BenchmarkCompareEQFunction-16    	167140707	         7.545 ns/op

goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkCompareEQInt64
BenchmarkCompareEQInt64-16       	206253358	         5.933 ns/op
BenchmarkCompareEQFloat
BenchmarkCompareEQFloat-16       	173147316	         5.936 ns/op
BenchmarkCompareEQString
BenchmarkCompareEQString-16      	134722820	         8.380 ns/op
BenchmarkCompareEQPString
BenchmarkCompareEQPString-16     	249906909	         4.567 ns/op
BenchmarkCompareEQBytes
BenchmarkCompareEQBytes-16       	132361765	         8.793 ns/op
BenchmarkCompareEQNil
BenchmarkCompareEQNil-16         	525513944	         2.351 ns/op
BenchmarkCompareNEInt
BenchmarkCompareNEInt-16         	204466880	         5.512 ns/op
BenchmarkCompareNEFloat
BenchmarkCompareNEFloat-16       	235253712	         5.576 ns/op
BenchmarkCompareNEString
BenchmarkCompareNEString-16      	129672235	         8.464 ns/op
BenchmarkCompareLTFloat
BenchmarkCompareLTFloat-16       	219071109	         4.984 ns/op
BenchmarkCompareLTString
BenchmarkCompareLTString-16      	140862020	         8.623 ns/op
BenchmarkCompareLTNil
BenchmarkCompareLTNil-16         	509819265	         2.447 ns/op
BenchmarkCompareEQFunction
BenchmarkCompareEQFunction-16    	191675306	         6.654 ns/op
PASS

**Review by dmitryax — APPROVED** [review]

**Review by kentquirk — APPROVED** [review]
Thank you for making these changes.

File: unreleased/fix-ottl-printf.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: pkg/ottl
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Remove a stray fmt.Printf, use a logger
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply unreleased/fix-ottl-printf.yaml:8]: @atoulme do you agree to apply this?

[the remark this answers is not in the document]
Comment (atoulme) [reply unreleased/fix-ottl-printf.yaml]: applied, thanks!

**Review by TylerHelmuth — APPROVED** [review]

**Review by evan-bradley — APPROVED** [review]

**Pull request merged**